solr/core/src/test/org/apache/solr/handler/component/QueryComponentPartialResultsTest.java

```diff
@@ -0,0 +1,216 @@
+import org.apache.solr.common.params.SolrParams;
+import org.apache.solr.common.util.NamedList;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.schema.*;
+import org.apache.solr.search.SortSpec;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+import java.util.*;
+
+public class QueryComponentPartialResultsTest extends SolrTestCaseJ4 {
+    private static int id = 0;
+    private static final String SORT_FIELD_NAME = "category";
+
+    @BeforeClass
+    public static void setup() {
+        assumeWorkingMockito();
+    }
+
+    @Test
+    public void includesPartialShardResultWhenUsingImplicitScoreSort() {
+        SortSpec sortSpec = MockSortSpecBuilder.create()
+                .withIncludesNonScoreOrDocSortField(false)
+                .build();
+        testPartialResultsForSortSpec(sortSpec, true);
+    }
+
+    @Test
+    public void includesPartialShardResultWhenUsingExplicitScoreSort() {
+        SortSpec sortSpec = MockSortSpecBuilder.create()
+                .withSortFields(new SortField[]{SortField.FIELD_SCORE})
+                .withIncludesNonScoreOrDocSortField(false)
+                .build();
+        testPartialResultsForSortSpec(sortSpec, true);
+    }
+
+    @Test
+    public void includesPartialShardResultWhenUsingExplicitDocSort() {
+        SortSpec sortSpec = MockSortSpecBuilder.create()
+                .withSortFields(new SortField[]{SortField.FIELD_DOC})
+                .withIncludesNonScoreOrDocSortField(false)
+                .build();
+        testPartialResultsForSortSpec(sortSpec, true);
+    }
+
+    @Test
+    public void excludesPartialShardResultWhenUsingNonScoreOrDocSortField() {
+        SortField sortField = new SortField(SORT_FIELD_NAME, SortField.Type.INT);
+        SortSpec sortSpec = MockSortSpecBuilder.create()
+                .withSortFields(new SortField[]{sortField})
+                .withIncludesNonScoreOrDocSortField(true)
+                .build();
+        testPartialResultsForSortSpec(sortSpec, false);
+    }
+
+    private void testPartialResultsForSortSpec(SortSpec sortSpec, boolean shouldIncludePartialShardResult) {
+        final int shard1Size = 2;
+        final int shard2Size = 3;
+
+        MockResponseBuilder responseBuilder = MockResponseBuilder.create().withSortSpec(sortSpec);
+
+        // shard 1 is marked partial
+        NamedList<Object> responseHeader1 = new NamedList<>();
+        responseHeader1.add(SolrQueryResponse.RESPONSE_HEADER_PARTIAL_RESULTS_KEY, Boolean.TRUE);
+
+        // shard 2 is not partial
+        NamedList<Object> responseHeader2 = new NamedList<>();
+
+        MockShardRequest shardRequest = MockShardRequest.create()
+                .withShardResponse(responseHeader1, createSolrDocumentList(shard1Size))
+                .withShardResponse(responseHeader2, createSolrDocumentList(shard2Size));
```

Comment:
Can this be pulled out into a field that only gets created once?

Reply:
Done.